### PR TITLE
Improve tiny model creation script

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

- Add the option to upload the created tiny models to Hub.
- Make tiny config corresponds better to the (reduced) tokenizer.
  - This gets quite complicated
  - But basically, just to make sure the `vocab_size` and `xxx_token_ids` in the tiny config correspond to what we have in the (reduced) tokenizer. 

Once this PR is approved, should I upload them to `hf-internal-testing`? (I remembered you said yes a few months ago, but want to be sure :-))

### CODE REVIEW

File: utils/create_dummy_models.py [hunk 1]
```diff
@@ -479,6 +505,55 @@ def fill_result_with_error(result, error, models_to_create):
             for model_arch in models_to_create[framework]:
                 result[framework][model_arch.__name__] = {"model": None, "checkpoint": None, "error": error}
 
+    result["processor"] = {type(p).__name__: p.__class__.__name__ for p in result["processor"]}
+
+
+def upload_model(model_dir, organization):
```
Comment (ydshieh) [inline utils/create_dummy_models.py:511]: For uploading

File: utils/create_dummy_models.py [hunk 2]
```diff
@@ -913,3 +1091,27 @@ def list_str(values):
 
     with open("simple_failed_report.txt", "w") as fp:
         fp.write(failed_report)
+
+    if args.upload:
```
Comment (ydshieh) [inline utils/create_dummy_models.py:1095]: For uploading

File: utils/create_dummy_models.py [hunk 3]
```diff
@@ -716,6 +884,9 @@ def build(config_class, models_to_create, output_dir):
     if result["warnings"]:
         logger.warning(result["warnings"])
 
+    # update `result["processor"]`
+    result["processor"] = {type(p).__name__: p.__class__.__name__ for p in processors}
```
Comment (ydshieh) [inline utils/create_dummy_models.py:888]: Just to include the created processors in the complete json report.

File: utils/create_dummy_models.py [hunk 4]
```diff
@@ -671,37 +833,43 @@ def build(config_class, models_to_create, output_dir):
...
 
-    # Update the config with the properties of the converted processors (smaller vocab size, image size, etc.)
-    for processor in processors:
-        if isinstance(processor, PreTrainedTokenizerBase):
-            vocab_size = processor.vocab_size
-            result["vocab_size"] = vocab_size
-    config_overrides = {k: v for k, v in result.items() if k in ["vocab_size"] and v is not None}
-    # Update `vocab_size`
+    try:
+        config_overrides = get_config_overrides(config_class, processors)
```
Comment (ydshieh) [inline utils/create_dummy_models.py:857]: Need a method as this gets more complicated

File: utils/create_dummy_models.py [hunk 5]
```diff
@@ -131,7 +139,7 @@ def get_config_class_from_processor_class(processor_class):
     """
 
     processor_prefix = processor_class.__name__
-    for postfix in ["Processor", "TokenizerFast", "Tokenizer", "FeatureExtractor"]:
+    for postfix in ["TokenizerFast", "Tokenizer", "ImageProcessor", "FeatureExtractor", "Processor"]:
```
Comment (ydshieh) [inline utils/create_dummy_models.py:142]: Need `ImageProcessor` after @amyeroberts 's work on Image Processors

File: utils/create_dummy_models.py [hunk 6]
```diff
@@ -248,12 +265,16 @@ def build_processor(config_class, processor_class):
...
+            raise ValueError(
+                f"`processor` (which is of type {processor.__class__.__name__}) should be an instance of"
+                f" {processor_class.__name__} or an Auto class!"
+            )
 
     return processor
 
 
-def get_tiny_config(config_class):
+def get_tiny_config(config_class, **model_tester_kwargs):
```
Comment (ydshieh) [inline utils/create_dummy_models.py:277]: To pass some arguments to model tester class's `__init__`, for example, `vocab_size`, `xxx_token_id` etc. See the explanation in `get_config_overrides`.

File: utils/create_dummy_models.py [hunk 7]
```diff
@@ -21,29 +21,37 @@
...
+from transformers.models.fsmt import configuration_fsmt
 from transformers.processing_utils import ProcessorMixin, transformers_module
 from transformers.tokenization_utils_base import PreTrainedTokenizerBase
 
 
+# make sure tokenizer plays nice with multiprocessing
+os.environ["TOKENIZERS_PARALLELISM"] = "false"
+
 logging.set_verbosity_error()
+logging.disable_progress_bar()
```
Comment (ydshieh) [inline utils/create_dummy_models.py:54]: As you mentioned in the previous PR - disable the progress bar.

File: utils/create_dummy_models.py [hunk 8]
```diff
@@ -620,6 +695,87 @@ def build_composite_models(config_class, output_dir):
...
+    if token is not None:
+        if isinstance(tokenizer, PreTrainedTokenizerFast):
+            token_id = tokenizer._convert_token_to_id_with_added_voc(token)
+        else:
+            token_id = tokenizer._convert_token_to_id(token)
+
+    return token_id
+
+
+def get_config_overrides(config_class, processors):
```
Comment (ydshieh) [inline utils/create_dummy_models.py:720]: So far (before this PR), we use the (converted) tokenizer's `vocab_size` to set `tiny_config.vocab_size`.

However, some model testers have `self.xxx_token_id = self.vocab_size - 1`, but we never use the updated `vocab_size` to recreate the model tester, so `self.xxx_token_id` was computed with  the original `vocab_size` (usually `99`).

Also, the special tokens in the tokenizer can have different ids before/after the conversion.

**This method tries its best effort to make sure the `xxx_token_id` in the tiny config (obtained from model tester) will correspond to the (converted) tokenizer, if the corresponding tokens exist in the tokenizer.**
(or at least, respect what has been specified in the model tester's `__init__`)

**Review by ydshieh — COMMENTED** [review]
Left a few comments to explain things.

File: utils/create_dummy_models.py [hunk 9]
```diff
@@ -479,6 +505,55 @@ def fill_result_with_error(result, error, models_to_create):
...
+        raise ValueError(error)
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+
+        repo = Repository(local_dir=tmpdir, clone_from=f"{organization}/{repo_name}")
+        repo.git_pull()
+        shutil.copytree(model_dir, tmpdir, dirs_exist_ok=True)
+
+        push = not repo_exist
+        if push:
```
Comment (LysandreJik) [inline utils/create_dummy_models.py:541]: nit: do we need this variable or can we just do `if repo_exist: else:`? :smile:

**Review by LysandreJik — APPROVED** [review]
Looks great to me! Great work @ydshieh!

File: utils/create_dummy_models.py (hunk 9 above)
Comment (ydshieh) [reply utils/create_dummy_models.py:541]: No need it, I will remove.

(I guess I was a bit tired at some point and found it's hard to understand `if not repo_exist` (negation) means `we want to push directly`, and didn't think of just changing the order 😄

**Pull request merged**